Try always generating navigation post title

tellthemachines opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Fixes #36593 and #36575.

Fully automates the Navigation post type naming process. When creating a new Navigation block, user doesn't have to give it a title or save it before starting to edit. It should get saved as part of the editor saving process.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

tellthemachines committed:
Try always generating navigation title

tellthemachines committed:
Show placeholder preview

talldan commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -100,18 +99,15 @@ export default function NavigationPlaceholder( {
 	const createNavigationMenu = useCallback(
-		async ( title = __( 'Untitled Navigation Menu' ), blocks = [] ) => {
+		async ( title, blocks = [] ) => {
+			if ( ! title ) {
+				return;
+			}
```
Testing and this will return `undefined` here because there's now no title. It looks like the way this is currently works in the PR is to trigger showing the `UnsavedInnerBlocks` component and have that create the `wp_navigation` post. I really think it'd be best to rely less on `UnsavedInnerBlocks` for that—that component was designed to handle a specific backwards compatibility edge case for patterns and not a whole lot else.

I feel like it'd be better to refactor or duplicate the code that generates a title in `UnsavedInnerBlocks` so that it can be used in the Placeholder too. That way there's always a title when we call `createNavigationMenu` and the change is a bit less involved.

It shouldn't be too much to achieve that. There's a few prequisites like calling the `useTemplatePartName` hook in the Placeholder component to get the area name. And also getting the entire list of menus so that the number added to the end of the name is correct. This bit of code below does the naming, it could be moved into a shared function or hook so that it can be used by both `UnsavedInnerBlocks` and `NavigationPlaceholder`:
https://github.com/WordPress/gutenberg/blob/88883a3ea23f5b3d8aff563f6a8447e4133b9e55/packages/block-library/src/navigation/edit/unsaved-inner-blocks.js#L137-L161

adamziel replied:
I added `useCreateNavigationMenu` that can be reused in both places. The process doesn't fully work yet and there is some flickering involved, but it's one step closer now. cc @tellthemachines

tellthemachines replied:
Thanks @adamziel ! I fixed all the bugs I could find and everything seems to be working correctly now. Would appreciate some extra testing 😄

adamziel committed:
Add useCreateNavigationMenu

adamziel commented on packages/block-library/src/navigation/edit/use-default-navigation-title.js:
```diff
@@ -0,0 +1,103 @@
+	// early.
+	const area = useTemplatePartAreaLabel( isDisabled ? undefined : clientId );
+
+	// Automatically save the uncontrolled blocks.
```
```suggestion
```

adamziel committed:
Update packages/block-library/src/navigation/edit/use-default-navigation-title.js

tellthemachines committed:
Fix placeholder and remove unecessary logic.

tellthemachines committed:
Pass clientId to placeholder.

tellthemachines commented:
All comments have been addressed and features are working as intended, so I'm marking this ready for review.

getdave reviewed (changes requested):
Thank you for all the work that's going into this. I gave it a spin and found the following issues:

### Blank Menu
Creating a new Navigation via `Select Menu -> Create New` results in a blank menu being created. 

* Delete all Navigations.
* Go to Site Editor and add Nav block.
* Click `Start empty` and don't add any blocks.
* No click `Select menu` again - notice there's a blank menu entry.
* Delete your Nav block.
* Re-add a new Nav block. Notice how `Select menu` appears.
* Notice how under `Select menu` there is a blank menu entry.
* Go to the `wp_navigation` Post screen. See that no Posts have been created. It appears this blank entry is a phantom.

Note this phantom entry persists even after you have added items to your Nav block and saved the resulting menu.

<img width="683" alt="Screenshot 2021-11-24 at 10 10 24" src="https://user-images.githubusercontent.com/444434/143218408-931df11f-fe75-4669-b25b-6c936866bcd0.png">


### Editing Menu in Trash

* Create a Nav block and add some items.
* Save.
* Go to `/wp-admin/edit.php?post_type=wp_navigation` and trash all your Navigations.
* Go back to the Site Editor.
* See that your Navigation items are still loaded into the Nav block even though they are in trash.
* Add a new item to the Navigation.
* Save.
* See how the item gets restored from Trash and is back to Publish state.
 

### Creating new Nav persists current Nav items

* Create a Navigation block and add some items.
* Save the Template so it is persisted.
* Go to Nav block and click `Select menu -> Create new`.
* Notice how a fresh Navigation Post is created but the same items are carried across from the current Navigation. 

Instead you should receive an entirely blank Navigation with no items.

getdave committed:
Reset inner blocks when creating new menu

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -356,6 +355,7 @@ function Navigation( {
 											onClose();
 										} }
 										onCreateNew={ () => {
+											replaceInnerBlocks( clientId, [] );
```
I added this because when you clicked "Create new menu" it would use the current inner blocks (i.e. from the current menu) and automatically create a new menu from those.

What we want is actually to remove all the inner blocks and reset in order that the placeholder will be shown.

getdave committed:
Only consider published Navigations as valid

getdave commented on packages/block-library/src/navigation/use-navigation-menu.js:
```diff
@@ -18,9 +18,17 @@ export default function useNavigationMenu( navigationMenuId ) {
+
+			// getEditedEntityRecord will return the post regardless of status.
+			// Therefore if the found post is not published then we should ignore it.
+			if ( navigationMenuId?.status !== 'publish' ) {
+				navigationMenu = null;
+			}
```
I don't think we should allow non-published Navigations to be used. Otherwise you can delete a Navigation and then accidentally have that restored simply by re-opening the Site Editor.

getdave committed:
Correct bug with checking status of Nav post

getdave committed:
UX to allow creating new menu if associated Nav post not found

Previously you could get into an error state if the Nav block referenced a Nav Post which had been deleted. You could not recover from this as the UI provided no way to "start over". This addition affords that ability.

adamziel committed:
Switch to useGenerateDefaultNavigationTitle

adamziel commented:
`useDefaultNavigationTitle` occasionally returned an empty string, e.g. when the dependencies were not available yet. In 74c49ad, I refactored it into `useGenerateDefaultNavigationTitle`. Now it doesn't just give up, but returns a promise that resolves with a proposed default title when the dependencies are loaded.

adamziel committed:
Wrap createNewMenu in useCallback

adamziel committed:
Rename createNewMenu to startWithEmptyMenu

adamziel reviewed (approved):
I think this works well 🚢 . There are now three different "create empty" buttons that do at least two different things:

<img width="616" alt="CleanShot 2021-11-24 at 15 26 28@2x" src="https://user-images.githubusercontent.com/205419/143261969-b0226793-eb3e-4163-b1e8-0e84ce2650d0.png">

Let's address that in a follow up PR. Great collaboration @getdave @tellthemachines !

talldan commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -306,6 +317,23 @@ function Navigation( {
 
+	// Show a warning if the selected menu is no longer available.
+	// TODO - the user should be able to select a new one?
+	if ( navigationMenuId && isNavigationMenuMissing ) {
+		return (
+			<div { ...blockProps }>
+				<Warning>
+					{ __(
+						'Navigation menu has been deleted or is unavailable. '
+					) }
+					<Button onClick={ startWithEmptyMenu } variant="link">
+						{ __( 'Create a new menu?' ) }
+					</Button>
+				</Warning>
+			</div>
+		);
+	}
```
This was removed in https://github.com/WordPress/gutenberg/pull/36507, I'm not sure why it made a reappearance here.

talldan commented on packages/block-library/src/navigation/use-navigation-menu.js:
```diff
@@ -22,9 +22,16 @@ export default function useNavigationMenu( navigationMenuId ) {
 				? getEditedEntityRecord( ...navigationMenuSingleArgs )
 				: null;
+
+			// getEditedEntityRecord will return the post regardless of status.
+			// Therefore if the found post is not published then we should ignore it.
+			if ( navigationMenu?.status !== 'publish' ) {
+				navigationMenu = null;
+			}
```
Navigation menus can be drafts too, that's an important part of the pattern insertion flow. Just wondering what the reasoning behind this change was?

talldan replied:
Related - https://github.com/WordPress/gutenberg/issues/37354
